src/Workspaces/Remote/ServiceHub/Shared/ServerDirectStream.cs

```diff
@@ -0,0 +1,101 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System;
+using System.IO;
+using System.IO.Pipes;
+using System.Runtime.Remoting;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace Microsoft.CodeAnalysis.Remote
+{
+    /// <summary>
+    /// Direct stream between service hub server and client to pass around big chunk of data
+    /// </summary>
+    internal class ServerDirectStream : Stream
```

Comment:
This seems liek a lot of complexity to support that.  I'd prefer you just hand out the pipe.  I don't see the need for the buffered stream, especially as PipeStreams can be given buffer sizes to begin with.